libc/string_mem.c

```diff
@@ -0,0 +1,396 @@
+	768, 257, 272, 306, 383, 384, 456, 511 };
+/* next values exceed 1 byte, (uint8_t)385 = -127, (uint8_t)511 = -1 and so on */
+
+
+TEST_GROUP(string_memset);
+TEST_GROUP(string_memmove);
+TEST_GROUP(string_memmove_big);
+
+
+TEST_SETUP(string_memset)
+{
+}
+
+
+TEST_TEAR_DOWN(string_memset)
+{
+}
+
+
+TEST(string_memset, basic)
+{
+	int i;
+	for (i = 0; i < VALUES_NUM; i++) {
+		ret = memset(buf, values[i], BUF_LEN);
+		TEST_ASSERT_EQUAL_PTR(buf, ret);
+		TEST_ASSERT_EACH_EQUAL_INT8((uint8_t)values[i], buf, BUF_LEN);
+	}
+}
+
+
+TEST(string_memset, zero)
+{
+	uint8_t val = 5;
+	ret = memset(&val, 0, 0);
+	TEST_ASSERT_EQUAL_PTR(&val, ret);
+	TEST_ASSERT_EQUAL_INT8(5, val);
+}
+
+
+TEST(string_memset, big)
+{
+	ret = memset(big_buf, 10, BIG_SIZE);
+	TEST_ASSERT_EQUAL_PTR(big_buf, ret);
+	TEST_ASSERT_EQUAL_MEMORY(expect_mem_big, big_buf, BIG_SIZE);
+}
+
+
+TEST(string_memset, single_byte)
+{
+	uint8_t val = 5;
+	ret = memset(&val, 0, 1);
+	TEST_ASSERT_EQUAL_PTR(&val, ret);
+	TEST_ASSERT_EQUAL_INT8(0, val);
+}
+
+
+TEST(string_memset, byte_val)
+{
+	int i;
+	for (i = 0; i < sizeof(buf_byte); i++) {
+		memset(buf_byte + i, i, 1);
+		buf_byte_exp[i] = i;
+	}
+	TEST_ASSERT_EQUAL_MEMORY(buf_byte_exp, buf_byte, sizeof(buf_byte));
+}
+
+
+TEST_SETUP(string_memmove)
+{
+	mem = (uint8_t *)calloc(MEM_SIZE, 1);
+	TEST_ASSERT_NOT_NULL(mem);
+	ret = memcpy(mem + OFFSET, TEST_STR, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(mem + OFFSET, ret);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST_TEAR_DOWN(string_memmove)
+{
+	free(mem);
+}
+
+
+TEST(string_memmove, no_overlap_end)
+{
+	uint8_t *ptr = mem + (MEM_SIZE - TEST_STR_SIZE);
+	/* Move to the end of allocated memory */
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+}
+
+
+TEST(string_memmove, no_overlap_between)
+{
+	int move_offset = OFFSET + TEST_STR_SIZE + 10;
+	uint8_t *ptr = mem + move_offset;
+	/* Move somewhere between */
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, move_offset - (OFFSET + TEST_STR_SIZE));
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + move_offset + TEST_STR_SIZE, MEM_SIZE - (move_offset + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, no_overlap_adjacent)
+{
+	uint8_t *ptr = mem + OFFSET + TEST_STR_SIZE;
+	/* Move so that memory blocks are next to each other*/
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
```

Comment:
the readability could be improved by using the same variable as src and dst params (clear relation between src and dst without even reading the comment), eg:

```c
uint8_t *src = mem + OFFSET;
ret = memmove(src + TEST_STR_SIZE, src, TEST_STR_SIZE);
```
please use the same approach in other tests

Reply:
I added dst pointer also so it is even more readable in my opinion.
```C
uint8_t *src = mem + OFFSET;
	uint8_t *dst = src + TEST_STR_SIZE;
	/* Move so that memory blocks are next to each other*/
	ret = memmove(dst, src, TEST_STR_SIZE);
```

Reply:
less readable, I need to analyze the values of `src` and `dst` to know the relation between them.

My version clearly showcases the relative offset between `src` and `dst` pointers  (and the len) by looking at a single line.